outlaw apl updates

Seny951 opened the pull request:
Outlaw APL update;

Some minor refactoring to align the cases more closely with SimC and simplify overcomplicated logic statements

Seny951 committed:
Merge branch 'dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
outlaw APL updates

Seny951 committed:
outlaw apl updates

mizone commented on HeroRotation_Rogue/Outlaw.lua:
Line 419 ~ 430

UseTrinkets

no code?

Seny951 replied:
Yes, for now this was intentional. I'm planning on coming back to it with a follow up PR.

From my testing It seems the current trinket code in the dragonflight branch doesn't work at all and never suggests any trinkets. I currently have no idea when this broke, so for now rather than fix old trinkets I've just removed it to get the APL updated.

Seny951 replied:
Updated PR: fixed trinkets

mizone replied:
why you keep ManicGrieftorch ?

whatever you have it or not , when meet HL.BossFilteredFightRemains("<=", 5) , it will show. please check it.

should we keep BladeFlurry always, even do ST ? would you make it like only recomend it only in combat?

mizone replied:
Line 329 S.CountTheOdds:IsAvailable()  

something wrong with this one

Seny951 replied:
Grieftorch is actually still good for prog when people are dying but it can be removed.

HL.BossFilteredFightRemains("<=", 5) is intentional and part of simC action, it's to dump the trinket if it's off CD at the end of an encounter:

use_item,name=manic_grieftorch,use_off_gcd=1,if=gcd.remains>gcd.max-0.1&!stealthed.all&buff.between_the_eyes.up|**fight_remains<=5**

Flurry is used pre pull only if UpperHanded Underhand is talented and it has to be used before stealth because otherwise it'll break it and if using UHUH it should be kept up ST all the time. 
Non uhuh builds still only use with 2+ targets.

What I will do though is put a stealth check on the pre pull so it doesn't suggest it if already stealthed (unless other flurry conditions are satisfied that are still valid flurry openers such as 2+ targets in range with subterfuge).

Seny951 replied:
Pushed the opener fix

Seny951 replied:
@mizone I see what you meant about the trinkets showing even if they aren't equipped. I fixed the trinket logic & also made blade flurry not suggest when out of combat & stealth if it's already up which will make the suggestions a little cleaner between trash pulls especially in m+

djokassss replied:
for me trinkets still dont show in rotation also shadow dance not showing up

Seny951 replied:
I'm going to need a little more information to investigate. 
Did you pull my fork branch?
how are you testing my changes?

 - To be clear if you're still testing against the official addon release or the dragonflight code branch this implementation hasn't gone into that yet.

Are you receiving any lua errors?
Are you max level?
are you using unusual talent choices?
What are your rogue & outlaw settings?

Just run a quick retest on a training dummy to make sure up-merges haven't broken anything and it still seems fine.

![image](https://github.com/herotc/hero-rotation/assets/19662794/bcfab467-878a-46e5-9f87-963dd677f871)
![image](https://github.com/herotc/hero-rotation/assets/19662794/c34b189f-9a4d-43a8-9a6a-da031850c2d8)

Seny951 committed:
fixed trinkets

Seny951 committed:
fixed trinkets

also fixed a bug where RtB was being suggested in certain circumstances when it was still on CD

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2.outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
fixed an issue with vanish conditions

 fixed an issue preventing vanish being used for Non HO builds

Seny951 committed:
fixed an issue with vanish conditions

 fixed an issue preventing vanish being used for Non HO builds

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Opener case for KiR builds to use ghostly strike instead of ambush

Seny951 committed:
Fixes for some inaccurate action conditions

mizone commented on HeroRotation_Rogue/Outlaw.lua:
Settings.Commons.OffGCDasOffGCD.Racials has been deleted in recent update

Seny951 committed:
blade flurry pre pull update

only suggest blade flurry for uhuh builds pre pull if not currently stealthed because it breaks it

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
bug fixes

prevent unequipped trinkets being suggested;
prevent blade flurry being suggested out of combat if it's already up;

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

framara commented:
for me current version of this fork is working very very good. Both KiR and HO builds, on both ST and keys. Thanks for the good work.

mizone commented:
I found another error.

Message: Interface/AddOns/HeroLib/Class/Spell/Main.lua:104: attempt to index local 'self' (a nil value)
Time: Tue Nov 14 00:02:55 2023
Count: 199
Stack: Interface/AddOns/HeroLib/Class/Spell/Main.lua:104: attempt to index local 'self' (a nil value)
[string "@Interface/AddOns/HeroLib/Class/Spell/Main.lua"]:104: in function IsAvailable' [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:329: in function <Interface/AddOns/HeroRotation_Rogue/Outlaw.lua:292> [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:394: in function <Interface/AddOns/HeroRotation_Rogue/Outlaw.lua:335> [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:691: in function ?'
[string "@Interface/AddOns/HeroRotation/Main.lua"]:456: in function <Interface/AddOns/HeroRotation/Main.lua:434>

Locals: self = nil
CheckPet = nil
(*temporary) = defined =[C]:-1
(*temporary) = nil
(*temporary) = nil
(*temporary) = "attempt to index local 'self' (a nil value)"
IsSpellKnown = defined =[C]:-1
IsPlayerSpell = defined =[C]:-1

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
fixed issue with bomb dispenser condition

Seny951 committed:
fixed typo in shadowmeld condition causing an error

Seny951 commented:
> I found another error.
> 
> Message: Interface/AddOns/HeroLib/Class/Spell/Main.lua:104: attempt to index local 'self' (a nil value) Time: Tue Nov 14 00:02:55 2023 Count: 199 Stack: Interface/AddOns/HeroLib/Class/Spell/Main.lua:104: attempt to index local 'self' (a nil value) [string "@Interface/AddOns/HeroLib/Class/Spell/Main.lua"]:104: in function IsAvailable' [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:329: in function <Interface/AddOns/HeroRotation_Rogue/Outlaw.lua:292> [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:394: in function <Interface/AddOns/HeroRotation_Rogue/Outlaw.lua:335> [string "@Interface/AddOns/HeroRotation_Rogue/Outlaw.lua"]:691: in function ?' [string "@Interface/AddOns/HeroRotation/Main.lua"]:456: in function <Interface/AddOns/HeroRotation/Main.lua:434>
> 
> Locals: self = nil CheckPet = nil (*temporary) = defined =[C]:-1 (*temporary) = nil (*temporary) = nil (*temporary) = "attempt to index local 'self' (a nil value)" IsSpellKnown = defined =[C]:-1 IsPlayerSpell = defined =[C]:-1

fixed

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -45,8 +45,15 @@ local I = Item.Rogue.Outlaw
+-- Trinkets
+local equip = Player:GetEquipment()
+local trinket1 = equip[13] and Item(equip[13]) or Item(0)
+local trinket2 = equip[14] and Item(equip[14]) or Item(0)
```
You will want a function to update this when equipment changes as well. See the Assassination module as an example.

```
local Equipment = Player:GetEquipment()
local TrinketItem1 = Equipment[13] and Item(Equipment[13]) or Item(0)
local TrinketItem2 = Equipment[14] and Item(Equipment[14]) or Item(0)

HL:RegisterForEvent(function()
  Equipment = Player:GetEquipment()
  TrinketItem1 = Equipment[13] and Item(Equipment[13]) or Item(0)
  TrinketItem2 = Equipment[14] and Item(Equipment[14]) or Item(0)
end, "PLAYER_EQUIPMENT_CHANGED" )

```

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -630,21 +687,16 @@ local function APL ()
-    if Stealthed_CtO() then
+
+    -- actions+=/call_action_list,name=stealth,if=stealthed.all
+    if Player:StealthUp(true, true) or Player:BuffUp(S.Shadowmeld) then
```
Not harmful, but `Player:StealthUp(true, true)` should already include Shadowmeld. (See StealthSpellsByType in Aura.lua)

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -269,204 +274,261 @@ local function Vanish_DPS_Condition ()
+	-- &!buff.greenskins_wickers.up
+	if S.BetweentheEyes:IsCastable() and Target:IsSpellInRange(S.BetweentheEyes) and not S.Crackshot:IsAvailable()
+		and (Player:BuffRemains(S.BetweentheEyes) < 4 or S.ImprovedBetweenTheEyes:IsAvailable() or S.GreenskinsWickers:IsAvailable()
+    or Player:HasTier(31, 4)) and Player:BuffDown(S.GreenskinsWickers) then
```
Should be checking T30 rather than T31 here I think?

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -269,204 +274,261 @@ local function Vanish_DPS_Condition ()
+  if CDsON() and S.Sepsis:IsAvailable() and S.Sepsis:IsReady() then
+    if S.Crackshot:IsAvailable() and S.BetweentheEyes:IsReady() and Finish_Condition() and not Player:StealthUp(true, true)
+      or not S.Crackshot:IsAvailable() and Target:FilteredTimeToDie(">", 11) and Player:BuffUp(S.BetweentheEyes) or HL.BossFilteredFightRemains("<", 11) then
+      if HR.cast(S.Sepsis, Settings.Outlaw.GCDasOffGCD.Sepsis) then return "Cast Sepsis" end
```
Think Lua is case-sensitive and won't like `cast` here?

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -269,204 +274,261 @@ local function Vanish_DPS_Condition ()
+  -- # Use Roll the Bones if reroll conditions are met, or just before buffs expire based on T31 and upcoming stealth cooldowns
+  -- actions.cds+=/roll_the_bones,if=variable.rtb_reroll|rtb_buffs.max_remains<=set_bonus.tier31_4pc+(cooldown.shadow_dance.remains<=1|cooldown.vanish.remains<=1)*6
+  if S.RolltheBones:IsReady() then
+    if RtB_Reroll() or Rogue.RtBRemains() <= num(Player:HasTier(31, 4)) + (num(S.ShadowDance:CooldownRemains() <=1) or num(S.Vanish:CooldownRemains() <= 1)) * 6 then
```
Rather than
`(num(S.ShadowDance:CooldownRemains() <=1) or num(S.Vanish:CooldownRemains() <= 1)) * 6`
this probably should be
`num(S.ShadowDance:CooldownRemains() <=1 or S.Vanish:CooldownRemains() <= 1) * 6`

EvanMichaels commented on HeroRotation_Rogue/Outlaw.lua:
```diff
@@ -269,204 +274,261 @@ local function Vanish_DPS_Condition ()
+  -- # Maintain Blade Flurry on 2+ targets, and on single target with Underhanded, or on cooldown at 5+ targets with Deft Maneuvers
+  -- actions.cds+=/blade_flurry,if=(spell_targets>=2-talent.underhanded_upper_hand&!stealthed.rogue)
+  -- &buff.blade_flurry.remains<gcd|talent.deft_maneuvers&spell_targets>=5&!variable.finish_condition
+  if S.BladeFlurry:IsReady() and (EnemiesBFCount >= 2 - num(S.UnderhandedUpperhand:IsAvailable()) and not Player:StealthUp(true, true))
```
stealthed.rogue should map to `Player:StealthUp(true, false)`

Seny951 committed:
Merge remote-tracking branch 'origin/dragonflight' into 10.2/outlaw

Seny951 committed:
addressed pr comments

Seny951 commented:
@EvanMichaels all raised issues fixed
